File: ui/windows/tests/Clambhook.Windows.Tests/DashboardStoreTests.cs
using System.Text.Json;
using Clambhook.Windows.Core;

namespace Clambhook.Windows.Tests;

public sealed class DashboardStoreTests
{
    [Fact]
    public async Task RefreshDashboardLoadsStatusProfilesAndServers()
    {
        var api = new FakeApi
        {
            Status = new StatusPayload(true, "A", [new ListenerStatusPayload("socks5", "127.0.0.1:1080", 3)]),
            Profiles = new ProfilesPayload(["A", "B"], "A"),
            Servers = new ServersPayload("A", [new ChainPayload("default", [new ServerPayload("london", "uk.example:443", "clambback")])]),
            Traffic = new TrafficSnapshotPayload
            {
                Summary = new TrafficSummaryPayload { ActiveConnections = 1, RxBps = 2048 },
                Connections = [new TrafficConnectionPayload { ConnId = "c1", State = "active", Target = "example.com:443" }]
            }
        };
        var store = new DashboardStore(api);

        await store.RefreshDashboardAsync();

        Assert.True(store.State.ApiOnline);
        Assert.True(store.State.Status.Running);
        Assert.Equal("A", store.State.ActiveProfile);
        Assert.Equal(3, store.State.ActiveConnections);
        Assert.Equal("london", store.State.Servers.Chains.Single().Servers.Single().Name);
        Assert.Equal("example.com:443", store.State.Traffic.Connections.Single().Target);
    }

    [Fact]
    public async Task RefreshDashboardStoresOfflineErrorOnFailure()
    {
        var store = new DashboardStore(new FakeApi { Error = new InvalidOperationException("boom") });

        await store.RefreshDashboardAsync();

        Assert.False(store.State.ApiOnline);
        Assert.Equal("boom", store.State.ErrorText);
    }

    [Fact]
    public async Task ActionsRefreshDashboardAfterSuccess()
    {
        var api = new FakeApi();
        var store = new DashboardStore(api);

        await store.ConnectAsync();
        await store.DisconnectAsync();
        await store.SetActiveProfileAsync("B");

        Assert.Equal(new[] { "connect", "disconnect", "profile:B" }, api.Actions);
        Assert.Equal(3, api.StatusCalls);
        Assert.Equal(3, api.ProfileCalls);
        Assert.Equal(3, api.ServerCalls);
        Assert.Equal(3, api.TrafficCalls);
    }

    [Fact]
    public void AppliesBandwidthAndLogEventsWithCaps()
    {
        var store = new DashboardStore(new FakeApi());

        store.ApplyEvent(new DaemonEvent(
            1,
            1,
            1,
            "connection.bytes",
            new Dictionary<string, JsonElement>
            {
                ["rx_delta"] = JsonSerializer.SerializeToElement(2048),
                ["tx_delta"] = JsonSerializer.SerializeToElement(1024),
                ["interval_ns"] = JsonSerializer.SerializeToElement(1_000_000_000)
            }));

        for (var i = 0; i < DashboardStore.MaxLogLines + 5; i++)
        {
            store.ApplyEvent(new DaemonEvent(
                0,
                (ulong)i,
                i,
                "log.line",
                new Dictionary<string, JsonElement> { ["line"] = JsonSerializer.SerializeToElement($"line-{i}") }));
        }

        Assert.Equal(2048, store.State.CurrentBandwidth.RxBps);
        Assert.Equal(1024, store.State.CurrentBandwidth.TxBps);
        Assert.Equal(DashboardStore.MaxLogLines, store.State.Logs.Count);
        Assert.Equal("line-5", store.State.Logs.First());
    }

    [Fact]
    public void DashboardViewStateBuildsRowsAndActions()
    {
        var state = new DashboardState
        {
            ApiOnline = true,
            Status = new StatusPayload(true, "A", [new ListenerStatusPayload("socks5", "127.0.0.1:1080", 2)]),
            Profiles = new ProfilesPayload(["A", "B"], "A"),
            Servers = new ServersPayload("A", [new ChainPayload("default", [new ServerPayload("london", "uk.example:443", "clambback")])]),
            Traffic = new TrafficSnapshotPayload
            {
                Summary = new TrafficSummaryPayload { ActiveConnections = 1, RxBps = 2048, TxBps = 1024 },
                Connections = [new TrafficConnectionPayload { State = "active", Target = "example.com:443", Network = "tcp", RxTotal = 2048, TxTotal = 1024 }]
            },
            Logs = ["started"]
        };

        var view = DashboardViewState.From(state, new DaemonSupervisor(), false, "");

        Assert.Equal("Running", view.StatusText);
        Assert.Equal("Disconnect", view.ConnectionActionText);
        Assert.True(view.CanToggleConnection);
        Assert.True(view.HasListeners);
        Assert.True(view.HasServers);
        Assert.True(view.HasTraffic);
        Assert.True(view.HasLogs);
        Assert.Equal("SOCKS5", view.Listeners.Single().Primary);
        Assert.Equal("example.com:443", view.Traffic.Single().Primary);
    }
}

internal sealed class FakeApi : IClambhookApi
{
    public StatusPayload Status { get; set; } = new();
    public ProfilesPayload Profiles { get; set; } = new(["A", "B"], "A");
    public ServersPayload Servers { get; set; } = new("A", []);
    public TrafficSnapshotPayload Traffic { get; set; } = new();
    public Exception? Error { get; set; }
    public List<string> Actions { get; } = [];
    public int StatusCalls { get; private set; }
    public int ProfileCalls { get; private set; }
    public int ServerCalls { get; private set; }
    public int TrafficCalls { get; private set; }

    public Task<StatusPayload> GetStatusAsync(CancellationToken cancellationToken = default)
    {
        StatusCalls++;
        if (Error is not null) throw Error;
        return Task.FromResult(Status);
    }

    public Task<ProfilesPayload> GetProfilesAsync(CancellationToken cancellationToken = default)
    {
        ProfileCalls++;
        if (Error is not null) throw Error;
        return Task.FromResult(Profiles);
    }

    public Task<ServersPayload> GetServersAsync(CancellationToken cancellationToken = default)
    {
        ServerCalls++;
        if (Error is not null) throw Error;
        return Task.FromResult(Servers);
    }

    public Task<TrafficSnapshotPayload> GetTrafficAsync(CancellationToken cancellationToken = default)
    {
        TrafficCalls++;
        if (Error is not null) throw Error;
        return Task.FromResult(Traffic);
    }

    public Task ConnectAsync(CancellationToken cancellationToken = default)
    {
        Actions.Add("connect");
        return Task.CompletedTask;
    }

    public Task DisconnectAsync(CancellationToken cancellationToken = default)
    {
        Actions.Add("disconnect");
        return Task.CompletedTask;
    }

    public Task SetActiveProfileAsync(string name, CancellationToken cancellationToken = default)
    {
        Actions.Add($"profile:{name}");
        return Task.CompletedTask;
    }
}

File: ui/apple/Tests/ClambhookSharedTests/DashboardStoreTests.swift
import XCTest
@testable import ClambhookShared

@MainActor
final class DashboardStoreTests: XCTestCase {
    func testRefreshLoadsDashboardAndPersistsWidgetSnapshot() async throws {
        let snapshotURL = temporaryURL("dashboard-snapshot.json")
        let snapshotStore = FileSnapshotStore(fileURL: snapshotURL)
        let api = FakeAPIClient()
        api.statusResult = StatusPayload(
            running: true,
            profile: "A",
            listeners: [ListenerStatusPayload(protocol: "socks5", addr: "127.0.0.1:1080", activeConns: 3)]
        )
        api.profilesResult = ProfilesPayload(profiles: ["A", "B"], active: "A")
        api.trafficResult = TrafficSnapshotPayload(
            summary: TrafficSummaryPayload(activeConnections: 1, rxBps: 2048, txBps: 1024),
            connections: [TrafficConnectionPayload(connID: "c1", state: "active", target: "example.com:443")]
        )
        api.serversResult = ServersPayload(
            profile: "A",
            chains: [ChainPayload(name: "default", servers: [
                ServerPayload(
                    name: "london",
                    address: "81.2.69.142:443",
                    protocol: "trojan",
                    geo: LocationPayload(country: "United Kingdom", countryCode: "GB", city: "London", latitude: 0, longitude: 0),
                    geoError: nil
                )
            ])]
        )
        let store = DashboardStore(api: api, snapshotStore: snapshotStore)

        await store.refreshDashboard()

        XCTAssertTrue(store.status.running)
        XCTAssertEqual(store.profiles.profiles, ["A", "B"])
        XCTAssertEqual(store.servers.chains.first?.servers.first?.name, "london")
        XCTAssertEqual(store.traffic.connections.first?.target, "example.com:443")
        let snapshot = try await snapshotStore.load()
        XCTAssertTrue(snapshot.apiOnline)
        XCTAssertTrue(snapshot.running)
        XCTAssertEqual(snapshot.profile, "A")
        XCTAssertEqual(snapshot.listenerCount, 1)
    }

    func testRefreshUsesSingleDashboardSnapshotWhenAvailable() async throws {
        let snapshotURL = temporaryURL("tunnel-dashboard-snapshot.json")
        let snapshotStore = FileSnapshotStore(fileURL: snapshotURL)
        let api = FakeDashboardAPI()
        api.dashboardResult = TunnelDashboardPayload(
            status: StatusPayload(
                running: true,
                profile: "phone",
                listeners: [ListenerStatusPayload(protocol: "tun", addr: "packet", activeConns: 2)]
            ),
            profiles: ProfilesPayload(profiles: ["phone", "backup"], active: "phone"),
            servers: ServersPayload(profile: "phone", chains: [
                ChainPayload(name: "proxy", servers: [
                    ServerPayload(name: "exit", address: "example.invalid:443", protocol: "shadowsocks")
                ])
            ]),
            rules: RulesPayload(profile: "phone", rules: [
                RulePayload(name: "ads", action: "block", domainSuffixes: ["ads.example.com"])
            ], generatedRules: [
                RulePayload(name: "subscription:ads:domains", action: "block", domainSuffixes: ["tracker.example.com"])
            ], effectiveRules: [
                RulePayload(name: "ads", action: "block", domainSuffixes: ["ads.example.com"]),
                RulePayload(name: "subscription:ads:domains", action: "block", domainSuffixes: ["tracker.example.com"])
            ]),
            policyGroups: PolicyGroupsPayload(profile: "phone", groups: [
                PolicyGroupPayload(name: "auto", type: "url-test", chains: ["proxy"], selectedChain: "proxy")
            ]),
            ruleSubscriptions: RuleSubscriptionsPayload(profile: "phone", subscriptions: [
                RuleSubscriptionPayload(
                    name: "ads",
                    url: "https://lists.example.invalid/ads.txt",
                    format: "auto",
                    action: "block",
                    cached: true,
                    domainCount: 1,
                    generatedRules: ["subscription:ads:domains"]
                )
            ]),
            traffic: TrafficSnapshotPayload(
                summary: TrafficSummaryPayload(activeConnections: 2, rxBps: 4096, txBps: 1024),
                connections: [TrafficConnectionPayload(connID: "c1", state: "active", target: "example.com:443")]
            ),
            dns: DNSPayload(
                profile: "phone",
                strategy: "encrypted",
                enabled: true,
                timeout: "5s",
                upstreams: [DNSUpstreamPayload(name: "cf", protocol: "doh", url: "https://cloudflare-dns.com/dns-query")],
                interceptsPort53: true,
                upstreamRoutes: [DNSUpstreamRoutePayload(name: "cf", protocol: "doh", target: "cloudflare-dns.com:443", network: "tcp", action: "chain", chainName: "proxy")]
            ),
            networkSettings: TunnelNetworkSettingsPayload(
                mtu: 1400,
                dnsServers: ["198.18.0.1"],
                includedRoutes: ["0.0.0.0/0"],
                excludedRoutes: ["127.0.0.0/8"]
            )
        )
        let store = DashboardStore(api: api, snapshotStore: snapshotStore)

        await store.refreshDashboard()

        XCTAssertEqual(api.dashboardCalls, 1)
        XCTAssertEqual(api.statusCalls, 0)
        XCTAssertEqual(api.profilesCalls, 0)
        XCTAssertEqual(api.serversCalls, 0)
        XCTAssertEqual(api.rulesCalls, 0)
        XCTAssertEqual(api.trafficCalls, 0)
        XCTAssertEqual(store.status.profile, "phone")
        XCTAssertEqual(store.profiles.profiles, ["phone", "backup"])
        XCTAssertEqual(store.servers.chains.first?.name, "proxy")
        XCTAssertEqual(store.rules.rules.first?.name, "ads")
        XCTAssertEqual(store.rules.generatedRules.first?.name, "subscription:ads:domains")
        XCTAssertEqual(store.rules.effectiveRules.count, 2)
        XCTAssertEqual(store.policyGroups.groups.first?.selectedChain, "proxy")
        XCTAssertEqual(store.ruleSubscriptions.subscriptions.first?.generatedRules, ["subscription:ads:domains"])
        XCTAssertEqual(store.traffic.summary.rxBps, 4096)
        XCTAssertEqual(store.dns.strategy, "encrypted")
        XCTAssertEqual(store.dns.upstreams.first?.name, "cf")
        XCTAssertEqual(store.dns.upstreamRoutes.first?.chainName, "proxy")
        XCTAssertEqual(store.networkSettings.dnsServers, ["198.18.0.1"])
        XCTAssertEqual(store.networkSettings.includedRoutes, ["0.0.0.0/0"])
        let snapshot = try await snapshotStore.load()
        XCTAssertTrue(snapshot.apiOnline)
        XCTAssertEqual(snapshot.profile, "phone")
        XCTAssertEqual(snapshot.listenerCount, 1)
    }

    func testDashboardPayloadDecodesMissingPolicyGroups() throws {
        let data = Data("""
        {
          "status": {"running": true, "profile": "A", "listeners": []},
          "profiles": {"profiles": ["A"], "active": "A"},
          "servers": {"profile": "A", "chains": []},
          "rules": {"profile": "A", "rules": []},
          "traffic": {"summary": {"active_connections": 0}, "connections": []}
        }
        """.utf8)

        let payload = try JSONDecoder().decode(TunnelDashboardPayload.self, from: data)

        XCTAssertEqual(payload.policyGroups, PolicyGroupsPayload())
        XCTAssertEqual(payload.ruleSubscriptions, RuleSubscriptionsPayload())
        XCTAssertEqual(payload.dns, DNSPayload())
        XCTAssertEqual(payload.networkSettings, TunnelNetworkSettingsPayload())
        XCTAssertEqual(payload.status.profile, "A")
    }

    func testDashboardPayloadDecodesDNSWhenPresent() throws {
        let data = Data("""
        {
          "status": {"running": true, "profile": "A", "listeners": []},
          "profiles": {"profiles": ["A"], "active": "A"},
          "servers": {"profile": "A", "chains": []},
          "rules": {"profile": "A", "rules": []},
          "traffic": {"summary": {"active_connections": 0}, "connections": []},
          "dns": {
            "profile": "A",
            "strategy": "encrypted",
            "enabled": true,
            "timeout": "5s",
            "intercepts_port_53": true,
            "upstreams": [{"name": "cf", "protocol": "doh", "url": "https://cloudflare-dns.com/dns-query"}],
            "upstream_routes": [{"name": "cf", "protocol": "doh", "target": "cloudflare-dns.com:443", "network": "tcp", "action": "chain", "chain_name": "proxy"}]
          }
        }
        """.utf8)

        let payload = try JSONDecoder().decode(TunnelDashboardPayload.self, from: data)

        XCTAssertTrue(payload.dns.enabled)
        XCTAssertEqual(payload.dns.strategy, "encrypted")
        XCTAssertEqual(payload.dns.upstreams.first?.targetDescription, "https://cloudflare-dns.com/dns-query")
        XCTAssertEqual(payload.dns.upstreamRoutes.first?.chainName, "proxy")
    }

    func testDashboardPayloadDecodesNetworkSettingsWhenPresent() throws {
        let data = Data("""
        {
          "status": {"running": true, "profile": "A", "listeners": []},
          "profiles": {"profiles": ["A"], "active": "A"},
          "servers": {"profile": "A", "chains": []},
          "rules": {"profile": "A", "rules": []},
          "traffic": {"summary": {"active_connections": 0}, "connections": []},
          "network_settings": {
            "mtu": 1400,
            "remote_address": "example.invalid",
            "ipv4": [{"address": "198.18.0.1", "prefix_len": 30}],
            "ipv6": [],
            "dns_servers": ["198.18.0.1"],
            "included_routes": ["0.0.0.0/0"],
            "excluded_routes": ["127.0.0.0/8"],
            "http_proxy": {"host": "127.0.0.1", "port": 18080}
          }
        }
        """.utf8)

        let payload = try JSONDecoder().decode(TunnelDashboardPayload.self, from: data)

        XCTAssertEqual(payload.networkSettings.mtu, 1400)
        XCTAssertEqual(payload.networkSettings.remoteAddress, "example.invalid")
        XCTAssertEqual(payload.networkSettings.ipv4.first?.address, "198.18.0.1")
        XCTAssertEqual(payload.networkSettings.dnsServers, ["198.18.0.1"])
        XCTAssertEqual(payload.networkSettings.includedRoutes, ["0.0.0.0/0"])
        XCTAssertEqual(payload.networkSettings.excludedRoutes, ["127.0.0.0/8"])
        XCTAssertEqual(payload.networkSettings.httpProxy, TunnelProxyPayload(host: "127.0.0.1", port: 18080))
    }

    func testTunnelPolicyGroupSelectionEditorUpdatesManualGroup() throws {
        let url = temporaryURL("policy-selection.toml")
        try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
        try """
        active = "A"

        [[profile]]
        name = "A"

          [[profile.policy_group]]
          name = "manual"
          type = "select"
          chains = ["proxy", "backup"]
          selected = "proxy"
        """.write(to: url, atomically: true, encoding: .utf8)

        try updateTunnelPolicyGroupSelection(
            configPath: url.path,
            profileName: "A",
            groupName: "manual",
            chainName: "backup"
        )

        let text = try String(contentsOf: url, encoding: .utf8)
        XCTAssertTrue(text.contains(#"selected = "backup""#))
    }

    func testApplyConnectionBytesKeepsLatestSamplesAndSnapshotRates() async throws {
        let snapshotURL = temporaryURL("bandwidth-snapshot.json")
        let snapshotStore = FileSnapshotStore(fileURL: snapshotURL)
        let store = DashboardStore(api: FakeAPIClient(), snapshotStore: snapshotStore)

        for index in 0..<65 {
            await store.apply(event: DaemonEvent(
                shardID: 1,
                lamport: UInt64(index + 1),
                tsNs: 0,
                type: "connection.bytes",
                data: [
                    "rx_delta": Double(index + 1) * 1024,
                    "tx_delta": Double(index + 1) * 512,
                    "interval_ns": Double(1_000_000_000),
                ]
            ))
        }

        XCTAssertEqual(store.bandwidthSamples.count, 60)
        XCTAssertEqual(store.bandwidthSamples.first?.rxBps, 6 * 1024)
        XCTAssertEqual(store.currentBandwidth.rxBps, 65 * 1024)
        let snapshot = try await snapshotStore.load()
        XCTAssertEqual(snapshot.rxBps, 65 * 1024)
        XCTAssertEqual(snapshot.txBps, 65 * 512)
    }

    func testApplyLogLineCapsRecentLogs() async {
        let store = DashboardStore(api: FakeAPIClient(), snapshotStore: .inMemory)

        for index in 0..<205 {
            await store.apply(event: DaemonEvent(
                shardID: 0,
                lamport: UInt64(index + 1),
                tsNs: 0,
                type: "log.line",
                data: ["line": "line-\(index)"]
            ))
        }

        XCTAssertEqual(store.logs.count, 200)
        XCTAssertEqual(store.logs.first, "line-5")
        XCTAssertEqual(store.logs.last, "line-204")
    }

    func testApplyLogLineUsesConfiguredRetention() async {
        let store = DashboardStore(api: FakeAPIClient(), snapshotStore: .inMemory, logRetention: 50)

        for index in 0..<55 {
            await store.apply(event: DaemonEvent(
                shardID: 0,
                lamport: UInt64(index + 1),
                tsNs: 0,
                type: "log.line",
                data: ["line": "line-\(index)"]
            ))
        }

        XCTAssertEqual(store.logs.count, 50)
        XCTAssertEqual(store.logs.first, "line-5")
        XCTAssertEqual(store.logs.last, "line-54")
    }

    func testSettingsNormalizationClampsValuesAndRejectsUnsupportedEndpoints() {
        let settings = AppSettings(
            apiEndpoint: URL(string: "ftp://example.test")!,
            daemonBinaryPath: " /tmp/clambhook \n",
            daemonConfigPath: "\t/tmp/config.toml ",
            refreshIntervalSeconds: 99,
            logRetention: 5,
            appGroupIdentifier: " "
        ).normalized()

        XCTAssertEqual(settings.apiEndpoint, defaultAPIEndpoint)
        XCTAssertEqual(settings.daemonBinaryPath, "/tmp/clambhook")
        XCTAssertEqual(settings.daemonConfigPath, "/tmp/config.toml")
        XCTAssertEqual(settings.refreshIntervalSeconds, maxRefreshIntervalSeconds)
        XCTAssertEqual(settings.logRetention, minLogRetention)
        XCTAssertEqual(settings.appGroupIdentifier, defaultAppGroupIdentifier)
        XCTAssertEqual(defaultAppGroupIdentifier, "group.org.jpfchang.clambhook")
        XCTAssertEqual(defaultPrivacyPolicyURL.absoluteString, "https://store.clambercloud.com/clambhook/privacy")
        XCTAssertEqual(defaultSupportURL.absoluteString, "https://store.clambercloud.com/clambhook/support")
    }

    func testCountryFlagAndRateFormatting() {
        XCTAssertEqual(countryFlag("GB"), "🇬🇧")
        XCTAssertEqual(countryFlag(""), "--")
        XCTAssertEqual(formatRate(500), "500 B/s")
        XCTAssertEqual(formatRate(1536), "1.5 KB/s")
        XCTAssertEqual(formatRate(2 * 1024 * 1024), "2.0 MB/s")
    }
}

private func temporaryURL(_ name: String) -> URL {
    FileManager.default.temporaryDirectory
        .appendingPathComponent(UUID().uuidString, isDirectory: true)
        .appendingPathComponent(name)
}

private class FakeAPIClient: ClambhookAPIProviding {
    var statusResult = StatusPayload(running: false, profile: "", listeners: [])
    var profilesResult = ProfilesPayload(profiles: [], active: "")
    var serversResult = ServersPayload(profile: "", chains: [])
    var policyGroupsResult = PolicyGroupsPayload()
    var rulesResult = RulesPayload(profile: "", rules: [])
    var trafficResult = TrafficSnapshotPayload()
    var dnsResult = DNSPayload()
    var ruleTestResult = RuleTestResponse()
    private(set) var connectCalls = 0
    private(set) var disconnectCalls = 0
    private(set) var selectedProfiles: [String] = []
    private(set) var selectedPolicyGroups: [(profile: String, group: String, chain: String)] = []
    private(set) var statusCalls = 0
    private(set) var profilesCalls = 0
    private(set) var serversCalls = 0
    private(set) var policyGroupCalls = 0
    private(set) var rulesCalls = 0
    private(set) var trafficCalls = 0
    private(set) var dnsCalls = 0

    func status() async throws -> StatusPayload {
        statusCalls += 1
        return statusResult
    }

    func profiles() async throws -> ProfilesPayload {
        profilesCalls += 1
        return profilesResult
    }

    func servers() async throws -> ServersPayload {
        serversCalls += 1
        return serversResult
    }

    func policyGroups() async throws -> PolicyGroupsPayload {
        policyGroupCalls += 1
        return policyGroupsResult
    }

    func rules() async throws -> RulesPayload {
        rulesCalls += 1
        return rulesResult
    }

    func dns() async throws -> DNSPayload {
        dnsCalls += 1
        return dnsResult
    }

    func testRule(network: String, target: String, profile: String) async throws -> RuleTestResponse {
        ruleTestResult
    }

    func traffic() async throws -> TrafficSnapshotPayload {
        trafficCalls += 1
        return trafficResult
    }

    func connect() async throws {
        connectCalls += 1
    }

    func disconnect() async throws {
        disconnectCalls += 1
    }

    func setActiveProfile(_ name: String) async throws {
        selectedProfiles.append(name)
    }

    func selectPolicyGroup(profile: String, group: String, chain: String) async throws -> PolicyGroupsPayload {
        selectedPolicyGroups.append((profile, group, chain))
        return policyGroupsResult
    }

    func testPolicyGroup(group: String, profile: String) async throws -> PolicyGroupsPayload {
        return policyGroupsResult
    }

    func updateDNS(_ request: DNSUpdateRequest, profile: String) async throws -> DNSPayload {
        return DNSPayload()
    }

    func exportConfig() async throws -> String { return "" }

    func importConfig(_ toml: String) async throws -> ConfigImportResponse {
        return ConfigImportResponse()
    }
}

private final class FakeDashboardAPI: FakeAPIClient, ClambhookDashboardProviding {
    var dashboardResult = TunnelDashboardPayload()
    private(set) var dashboardCalls = 0

    func dashboard() async throws -> TunnelDashboardPayload {
        dashboardCalls += 1
        return dashboardResult
    }
}
